# From: hkfuertes <hkfuertes@example.com>
# Date: Sun, 12 Oct 2025 00:18:00 +0200
# Subject: [PATCH] arm64: dts: qcom: Add MF68E MiFi router with ST7735S SPI display
# 
# Add device tree for MF68E 4G LTE MiFi router based on MSM8916.
# Includes ST7735S 128x128 SPI display panel configuration.
# 
# Signed-off-by: hkfuertes <hkfuertes@example.com>

--- /dev/null
+++ b/arch/arm64/boot/dts/qcom/msm8916-generic-mf68e.dts
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/dts-v1/;
+
+#include "msm8916-ufi.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	model = "MF68E 4G LTE Mobile WiFi Router";
+	compatible = "generic,mf68e", "qcom,msm8916";
+
+	bat: battery {
+		compatible = "simple-battery";
+		voltage-min-design-microvolt = <3400000>;
+		voltage-max-design-microvolt = <4350000>;
+		energy-full-design-microwatt-hours = <11100000>;
+		charge-full-design-microamp-hours = <3000000>;
+
+		ocv-capacity-celsius = <25>;
+		ocv-capacity-table-0 = <4330000 100>, <4265000 95>,
+			<4208000 90>, <4154000 85>, <4102000 80>,
+			<4063000 75>, <4006000 70>, <3972000 65>,
+			<3931000 60>, <3880000 55>, <3841000 50>,
+			<3817000 45>, <3798000 40>, <3783000 35>,
+			<3767000 30>, <3747000 25>, <3729000 20>,
+			<3700000 16>, <3687000 13>, <3683000 11>,
+			<3681000 10>, <3680000 9>, <3678000 8>,
+			<3674000 7>, <3666000 6>, <3641000 5>,
+			<3597000 4>, <3537000 3>, <3457000 2>,
+			<3336000 1>, <3000000 0>;
+	};
+
+	usb_id: usb-id {
+		compatible = "linux,extcon-usb-gpio";
+		id-gpios = <&tlmm 110 GPIO_ACTIVE_HIGH>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&usb_id_default>;
+	};
+
+	backlight: backlight {
+		compatible = "gpio-backlight";
+		gpios = <&tlmm 98 GPIO_ACTIVE_HIGH>;
+		default-on;
+	};
+};
+
+&pm8916_bms {
+	status = "okay";
+	monitored-battery = <&bat>;
+	power-supplies = <&pm8916_charger>;
+};
+
+&pm8916_charger {
+	status = "okay";
+	monitored-battery = <&bat>;
+	qcom,fast-charge-safe-current = <900000>;
+	qcom,fast-charge-safe-voltage = <4350000>;
+};
+
+/* ST7735S 128x128 SPI Display Panel */
+&blsp_spi5 {
+	status = "okay";
+	
+	panel@0 {
+		compatible = "sitronix,st7735r";
+		reg = <0>;
+		spi-max-frequency = <16000000>;
+		
+		reset-gpios = <&tlmm 25 GPIO_ACTIVE_LOW>;
+		dc-gpios = <&tlmm 24 GPIO_ACTIVE_HIGH>;
+		backlight = <&backlight>;
+		
+		width = <128>;
+		height = <128>;
+		rotation = <0>;
+		
+		vdd-supply = <&pm8916_l17>;
+		vddio-supply = <&pm8916_l6>;
+		
+		pinctrl-names = "default", "sleep";
+		pinctrl-0 = <&panel_default>;
+		pinctrl-1 = <&panel_sleep>;
+	};
+};
+
+/* Define L17 regulator for display VDD (2.85V) */
+&rpm_requests {
+	pm8916-regulators {
+		pm8916_l17: l17 {
+			regulator-min-microvolt = <2850000>;
+			regulator-max-microvolt = <2850000>;
+		};
+	};
+};
+
+&usb {
+	status = "okay";
+	extcon = <&usb_id>, <&usb_id>;
+};
+
+&usb_hs_phy {
+	extcon = <&usb_id>;
+};
+
+&tlmm {
+	usb_id_default: usb-id-default-state {
+		pins = "gpio110";
+		function = "gpio";
+		drive-strength = <8>;
+		bias-pull-up;
+	};
+	
+	panel_default: panel-default-state {
+		reset-pins {
+			pins = "gpio25";
+			function = "gpio";
+			drive-strength = <8>;
+			bias-disable;
+			output-high;
+		};
+		
+		dc-pins {
+			pins = "gpio24";
+			function = "gpio";
+			drive-strength = <8>;
+			bias-disable;
+		};
+		
+		enable-pins {
+			pins = "gpio97";
+			function = "gpio";
+			drive-strength = <8>;
+			bias-disable;
+			output-high;
+		};
+	};
+	
+	panel_sleep: panel-sleep-state {
+		pins = "gpio24", "gpio25", "gpio97";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-pull-down;
+		output-low;
+	};
+};
